blog/content/post/2021-02-05-testing.md

```diff
@@ -0,0 +1,134 @@
+bring the expected results.
+
+![testing_chart](/img/test1.png)
+
+Our validation infrastructure in 3mdeb consists of ~150 tests. Performing the
+whole testing process step by step would take days, especially that the whole
+process is repetitive. Summarizing, manual testing in this case costs more and
+slows down the whole development pace, when great delay occurs between code
+release and feedback for the programmers. If we would still try to test
+manually, our knowledge of code quality would be worse and the delivery would be
+delayed.
+
+The whole process may be performed in a much more convenient way. Tests can be
+run by the developer, who needs only a brief information about the testing
+environment. Clear documentation and ready-to-use physical platforms with system
+of automatic availability management make it possible to run both, the single
+tests and complete suites, concerning a set of functionalities that must be
+checked. In result we receive immediate feedback about the changes made and the
+features that were added.
+
+#### Remote Testing Environment
+
+As we are performing our test on hardware we had to found a way to automate
+management and operations on physical platforms. To meet that goal we have
+created Remote Testing Environment - a Orange Pi Zero hat that gives us
+ability to remotely control power, simulate pressing of buttons like power on
+or reset, read serial output and flash firmware using physical cable connection
+that can unbrick device without any manual intervention. This setup has a
+corresponding configuration file for every platform model which allow for
+smooth remote test execution. You can find [RTE](https://jira-3mdeb.atlassian.net/browse/SHA-195?focusedCommentId=55760&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-55760) in our shop.
+
+![RTA](/img/RTA.png)
+
+An automated testing infrastructure can be paired with constant integration and
+delivery pipeline, to enhance the overall performance. We successfully use such
+approach when testing our projects. It all begins when most recent changes for
+the release are ready in our core repository and marked as a release candidate.
+From this point, the automatic process of building usable binaries begins,
+signing it with authenticity proving keys, testing, and delivering the final
+product to the customer, as an online ready-to-download file. Because these
+time-consuming tasks are performing automatically, it leaves developers more
+time for truly creative work on delivering the highest quality of product.
+
+Test verification process needs understanding of the tested functionality and
+further interpretation of the results. In case of the automated testing, the
+output provides result and definition for getting the necessary information for
+users, clients and developers. When the complete regression testing results for
+the given release are ready, the next step is to make them accessible and easy
+to understand.
+
```

Comment:
@KBanecka please add also note about the RTE tag on our blog: https://blog.3mdeb.com/tags/rte/

Reply:
done